Error-Messages.md

```diff
@@ -309,4 +316,226 @@ Integer i1 = null;
+    return foo.toString();
+  }
+
+  //class field "foo" is non-null at call site of "requiresFoo", hence no error will be reported.
+  void test1(){ 
+    foo = new Object();
+    requiresFoo()
+  }   
+
+  //class field "foo" is non-null via call to "ensuresNonnullFoo" at call site of "requiresFoo", hence no error will be reported.
+  void test2(){ 
+    ensuresNonnullFoo()
+    requiresFoo()
+  }
+
+  //class field "foo" is not non-null at call site of "requiresFoo", hence the above error will be reported.
+  void test3(){
+    requiresFoo()
+  }
+}
+```
+
+### empty @RequiresNonnull/@EnsuresNonnull is the default precondition for every method, please remove it
+Empty `@RequiresNonnull` and `@EnsuresNonnull` is the default type for all methods. Therefore we report an error on such cases to keep the code simple.
+
+Please see the example below where this error will be reported.
+```java
+class C {
+
+  Object foo;
+  
+  @EnsuresNonnull()
+  void ensures(){
+
+  }
+
+  @RequiresNonnull()
+  void requires(){
+
+  }
+}
+```
+
+### currently @RequiresNonnull/@EnsuresNonnull supports only class fields of the method receiver: fieldName is not supported
+This error indicates that the annotation parameter is violating the following syntax rule:
+```
+The receiver of selected fields in annotation can only be the receiver of the method
+```
+Please see the example below where this error will be reported.
+```java
+class C {
+
+  Object foo;
+  C other;
+  
+  @EnsuresNonnull("other.foo") // error: can only be either "foo" or "this.foo"
+  void ensures(){
+
+  }
+
+  @RequiresNonnull("other.foo") // error: can only be either "foo" or "this.foo"
+  void requires(){
+
+  }
+}
+```
+
+### for @RequiresNonnull/@EnsuresNonnull annotation, cannot find instance field in class
+This error indicates that the annotation parameter is violating the following syntax rule:
+```
+All parameters given in the annotation must be one of the fields of the class or its super classes.
+```
+Please see the example below where this error will be reported.
+
+```java
+class Super {
+  Object bar;
+}
+
+class C extends Super{
+
+  Object foo;
+  
+  @EnsuresNonnull("foo") // no error: field is present in "C"
+  void ensures(){
+    foo = new Object();
+  }
+
+  @RequiresNonnull("bar") // no error: field is present in super class "Super"
+  void requires(){
+    //not important
+  }
+
+  @RequiresNonnull("tmp") // error: field is not present in this class or any of it's super classes.
+  void requires(){
+    //not important
+  }
+}
+```
+
+### postcondition inheritance is violated, this method must guarantee that all fields written in the @EnsuresNonNull annotation of overridden method are @NonNull at exit point as well. Fields must explicitly appear as parameters at this method @EnsuresNonNull annotation
+
+This error indicates that the annotation violating the following inheritance rule:
+```
+Every method must satisfy all postcondition of its super method.
```

Comment:
A few of this documentation sections are very much written from the point of view of tool implementation, rather than keeping in mind what developers not familiar with NullAway should be taking out from this. Instead of just listing the inheritance rule (which is fine, but not sufficient), we should probably have a small paragraph explaining why this is important:

> "Without this rule, a postcondition of `@EnsuresNonNull("field") void foo(...)` on type `Supertype` might be violated for `o.foo(...)` on `Supertype o = new Subtype()`, depending on whether and how `Subtype` overrides the method `foo` of `Supertype`."

(There might be a clearer way to write it, the point is to make the developer aware of what could potentially break in their code if this rule wasn't there)

We could possibly just link to https://github.com/uber/NullAway/wiki/Error-Messages#method-returns-nullable-but-superclass-method-returns-nonnull too.